### PR TITLE
DEPR: observed=False default in groupby

### LINKED ISSUE #30552: PERF: groupby with many empty groups memory blowup
Suppose we have a Categorical with many unused categories:

```
cat = pd.Categorical(range(24), categories=range(10**5))

df = pd.DataFrame({"A": cat, "B": range(24), "C": range(24), "D": 1})

gb = df.groupby(["A", "B", "C"])

>>> gb.size()  # memory balloons to 9+ GB before i kill it
```

There are only 24 rows in this DataFrame, so we shouldn't be creating millions of groups.

Without the Categorical, but just a large cross-product that implies many empty groups, this works fine:

```
df = pd.DataFrame({n: range(12) for n in range(8)})

gb = df.groupby(list(range(7)))
gb.size() # <-- works fine
```

### LINKED ISSUE #43999: ENH: Default to observed=True in `DataFrame.groupby`
#### Is your feature request related to a problem?

The default behaviour of `pandas.DataFrame.groupby` is currently different depending on the type of the groupers (when one of the groupers is categorical, unobserved categories are added to the groupby by default). 

I feel like making the groupby API  be consistent by default and regardless of the underlying data type would provide a better user experience. 

#### Describe the solution you'd like

Default to `observed=False` in `pandas.DataFrame.groupby`.

#### API breaking implications

Would break backwards-compatibility.

#### Describe alternatives you've considered

So far the only option I can think of is to add `observed=True`  to every groupby I write to make sure it will behave correctly no matter what kind of data gets passed to it.

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #43999 (Replace xxxx with the GitHub issue number)
- [x] closes #30552
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by topper-123** [comment]
for users who work interaactively (ipython, jupyter,..), Is there way to avoid the `FutureWarning` coming repeatedly? It will probably get quite annoying after a while...

Otherwise LGTM.

**Comment by rhshadrach** [comment]
Thanks @topper-123 

> for users who work interaactively (ipython, jupyter,..), Is there way to avoid the FutureWarning coming repeatedly? It will probably get quite annoying after a while...

Users can specify `observed=True|False`. Also, just to be sure, this will only warn if one of the groupings is categorical.

I think you're thinking of some global option? I would be -1 on that, and hope this will be rare enough that it won't pester users. They'll need to specify `observed=True` if they want to adopt future behavior or `observed=False` if they want to maintain current behavior eventually.

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -93,10 +93,10 @@ Other API changes
 Deprecations
 ~~~~~~~~~~~~
 - Deprecating pinning ``group.name`` to each group in :meth:`SeriesGroupBy.aggregate` aggregations; if your operation requires utilizing the groupby keys, iterate over the groupby object instead (:issue:`41090`)
+- Deprecated the default of ``observed=False`` in :meth:`DataFrame.groupby` and :meth:`Series.groupby` (:issue:`43999`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:96]: Could you mention this will be `True` in the future?

Comment (rhshadrach) [reply doc/source/whatsnew/v2.1.0.rst]: Good call - done.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @rhshadrach

**Pull request merged**